packages-exp/app-check-exp/src/internal-api.test.ts

```diff
@@ -0,0 +1,443 @@
+/**
+ * @license
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import '../test/setup';
+import { expect } from 'chai';
+import { SinonStub, spy, stub, useFakeTimers } from 'sinon';
+import { deleteApp, FirebaseApp } from '@firebase/app-exp';
+import {
+  FAKE_SITE_KEY,
+  getFullApp,
+  getFakeCustomTokenProvider,
+  getFakePlatformLoggingProvider,
+  removegreCAPTCHAScriptsOnPage
+} from '../test/util';
+import { initializeAppCheck } from './api';
+import {
+  getToken,
+  addTokenListener,
+  removeTokenListener,
+  formatDummyToken,
+  defaultTokenErrorData
+} from './internal-api';
+import * as reCAPTCHA from './recaptcha';
+import * as client from './client';
+import * as storage from './storage';
+import { getState, clearState, setState, getDebugState } from './state';
+import { AppCheckTokenListener } from '../src/types';
+import { Deferred } from '@firebase/util';
+import { ReCaptchaV3Provider } from './providers';
+
+const fakePlatformLoggingProvider = getFakePlatformLoggingProvider();
+
+describe('internal api', () => {
+  let app: FirebaseApp;
+
+  beforeEach(() => {
+    app = getFullApp();
+  });
+
+  afterEach(() => {
+    clearState();
+    removegreCAPTCHAScriptsOnPage();
+    return deleteApp(app);
+  });
+  // TODO: test error conditions
+  describe('getToken()', () => {
+    const fakeRecaptchaToken = 'fake-recaptcha-token';
+    const fakeRecaptchaAppCheckToken = {
+      token: 'fake-recaptcha-app-check-token',
+      expireTimeMillis: 123,
+      issuedAtTimeMillis: 0
+    };
+
+    const fakeCachedAppCheckToken = {
+      token: 'fake-cached-app-check-token',
+      expireTimeMillis: 123,
+      issuedAtTimeMillis: 0
+    };
+
+    it('uses customTokenProvider to get an AppCheck token', async () => {
+      const clock = useFakeTimers();
+      const customTokenProvider = getFakeCustomTokenProvider();
+      const customProviderSpy = spy(customTokenProvider, 'getToken');
+
+      initializeAppCheck(app, { provider: customTokenProvider });
+      const token = await getToken(app, fakePlatformLoggingProvider);
+
+      expect(customProviderSpy).to.be.called;
+      expect(token).to.deep.equal({
+        token: 'fake-custom-app-check-token'
+      });
+
+      clock.restore();
+    });
+
+    it('uses reCAPTCHA token to exchange for AppCheck token if no customTokenProvider is provided', async () => {
```

Comment:
The test name is no longer appropriate for the Provider approach we are taking. Perhaps simplify it to something like: `uses RecaptchaV3Provider to get an AppCheck token` similar to the test name for `customTokenProvider`.

Reply:
Done.